libraries/AP_GyroFFT/AP_GyroFFT.cpp

```diff
@@ -0,0 +1,654 @@
+    // @Range: 0 0.9
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("WINDOW_OLAP", 6, AP_GyroFFT, _window_overlap, FFT_DEFAULT_WINDOW_OVERLAP),
+
+    // @Param: FREQ_HOVER
+    // @DisplayName: FFT learned hover frequency
+    // @Description: The learned hover noise frequency
+    // @Range: 0 250
+    // @User: Advanced
+    AP_GROUPINFO("FREQ_HOVER", 7, AP_GyroFFT, _freq_hover_hz, 80.0f),
+
+    // @Param: THR_REF
+    // @DisplayName: FFT learned thrust reference
+    // @Description: FFT learned thrust reference for the hover frequency and FFT minimum frequency.
+    // @Range: 0.01 0.9
+    // @User: Advanced
+    AP_GROUPINFO("THR_REF", 8, AP_GyroFFT, _throttle_ref, FFT_THR_REF_DEFAULT),
+
+    // @Param: SNR_REF
+    // @DisplayName: FFT SNR reference threshold
+    // @Description: FFT SNR reference threshold in dB at which a signal is determined to be present.
+    // @Range: 0.0 100.0
+    // @User: Advanced
+    AP_GROUPINFO("SNR_REF", 9, AP_GyroFFT, _snr_threshold_db, FFT_SNR_DEFAULT),
+
+    // @Param: ATT_REF
+    // @DisplayName: FFT attenuation for bandwidth
+    // @Description: FFT desired attenuation level in dB for bandwidth calculation. The bandwidth is calculated by comparing peak power output with the attenuated version.
+    // @Range: 0 100
+    // @User: Advanced
+    AP_GROUPINFO("ATT_REF", 10, AP_GyroFFT, _attenuation_power_db, 15),
+
+    // @Param: BW_HOVER
+    // @DisplayName: FFT learned bandwidth at hover
+    // @Description: FFT learned bandwidth at hover for the attenuation frequencies.
+    // @Range: 0 200
+    // @User: Advanced
+    AP_GROUPINFO("BW_HOVER",11, AP_GyroFFT, _bandwidth_hover_hz, 20),
+
+    AP_GROUPEND
+};
+
+// The FFT splits the frequency domain into an number of bins
+// A sampling frequency of 1000 and max frequency (Nyquist) of 500 at a window size of 32 gives 16 frequency bins each 31.25Hz wide
+// The first bin is used to store the DC and Nyquist values combined.
+// Eg [DC/Nyquist], [16,47), [47,78), [78,109) etc
+// For a loop rate of 800Hz, 16 bins each 25Hz wide
+// Eg X[0]=[DC/Nyquist], X[1]=[12,37), X[2]=[37,62), X[3]=[62,87), X[4]=[87,112)
+// So middle frequency is X[n] * 25 and the range is X[n] * 25 - 12 < f < X[n] * 25 + 12
+
+// Maximum tolerated number of cycles with missing signal
+#define FFT_MAX_MISSED_UPDATES 3
+
+const extern AP_HAL::HAL& hal;
+
+AP_GyroFFT::AP_GyroFFT()
+{
+    _multiplier = 2000.0f / radians(2000); // scale from radian gyro output back to degrees
+    _thread_state._noise_needs_calibration = 0x07; // all axes need calibration
+    AP_Param::setup_object_defaults(this, var_info);
+
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_GyroFFT must be singleton");
+    }
+    _singleton = this;
+}
+
+// initialize the FFT parameters and engine
+void AP_GyroFFT::init(uint32_t target_looptime_us, AP_InertialSensor& ins)
```

Comment:
we can get AP_InertialSensor as a singleton

Reply:
ok

Reply:
Gone